integration-tests/src/test/java/io/opentelemetry/B3PropagationIntegrationTest.java

```diff
@@ -0,0 +1,245 @@
+import io.opentelemetry.context.propagation.TextMapGetter;
+import io.opentelemetry.context.propagation.TextMapPropagator;
+import io.opentelemetry.extension.trace.propagation.B3Propagator;
+import io.opentelemetry.sdk.OpenTelemetrySdk;
+import io.opentelemetry.sdk.testing.exporter.InMemorySpanExporter;
+import io.opentelemetry.sdk.trace.SdkTracerProvider;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.sdk.trace.export.SimpleSpanProcessor;
+import java.io.IOException;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Stream;
+import org.jetbrains.annotations.Nullable;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.testcontainers.shaded.okhttp3.Call;
+import org.testcontainers.shaded.okhttp3.OkHttpClient;
+import org.testcontainers.shaded.okhttp3.Request;
+import org.testcontainers.shaded.okhttp3.Response;
+import org.testcontainers.shaded.okhttp3.ResponseBody;
+import spark.Service;
+
+/** Integration tests for the B3 propagators, in various configurations. */
+public class B3PropagationIntegrationTest {
+
+  private static final int server1Port = new Random().nextInt(40000) + 1024;
+  private static final int server2Port = new Random().nextInt(40000) + 1024;
+  private static final InMemorySpanExporter spanExporter = InMemorySpanExporter.create();
+
+  private OpenTelemetrySdk clientSdk;
+  private OpenTelemetrySdk server1Sdk;
+  private OpenTelemetrySdk server2Sdk;
+  private Service server1;
+  private Service server2;
+
+  @SuppressWarnings("UnusedMethod")
+  private static Stream<TextMapPropagator> propagators() {
+    return Stream.of(B3Propagator.injectingMultiHeaders(), B3Propagator.injectingSingleHeader());
+  }
+
+  void setup(TextMapPropagator propagator) {
+    setupClient(propagator);
+    setupServer2(propagator);
+    setupServer1(propagator);
+  }
+
+  @AfterEach
+  void shutdown() {
+    if (server2 != null) {
+      server2.stop();
+    }
+    if (server1 != null) {
+      server1.stop();
+    }
+    spanExporter.reset();
+  }
+
+  @ParameterizedTest
+  @MethodSource("propagators")
+  void propagation(TextMapPropagator propagator) throws IOException {
+    setup(propagator);
+    OkHttpClient httpClient = createPropagatingClient(clientSdk);
+
+    Span span = clientSdk.getTracer("testTracer").spanBuilder("clientSpan").startSpan();
+    try (Scope ignored = span.makeCurrent()) {
+      Call call =
+          httpClient.newCall(
+              new Request.Builder().get().url("http://localhost:" + server1Port + "/test").build());
+
+      try (Response r = call.execute()) {
+        ResponseBody body = r.body();
+        assertThat(body.string()).isEqualTo("OK");
+      }
+    } finally {
+      span.end();
+    }
+
+    List<SpanData> finishedSpanItems = spanExporter.getFinishedSpanItems();
+    assertThat(finishedSpanItems).hasSize(3);
```

Comment:
Using a static int variable (with a descriptive name) might make more sense here - otherwise it is unclear why # span items should be `3`

Reply:
I can add a comment. I don't think having a static variable would help with clarity here.

Reply:
comments added. good suggestion